src/lib/__tests__/index-test.ts

```diff
@@ -0,0 +1,855 @@
+          instantSearchInstance,
+        })
+      );
+
+      expect(searchBox.render).toHaveBeenCalledTimes(1);
+      expect(pagination.render).toHaveBeenCalledTimes(1);
+      expect(pagination.render).toHaveBeenCalledWith({
+        instantSearchInstance,
+        results: expect.any(algoliasearchHelper.SearchResults),
+        state: expect.any(algoliasearchHelper.SearchParameters),
+        helper: instance.getHelper(),
+        templatesConfig: instantSearchInstance.templatesConfig,
+        createURL: instantSearchInstance._createAbsoluteURL,
+        searchMetadata: {
+          isSearchStalled: instantSearchInstance._isSearchStalled,
+        },
+      });
+    });
+  });
+
+  describe('dispose', () => {
+    it('calls `dispose` on its widgets', () => {
+      const instance = index({ indexName: 'index_name' });
+      const instantSearchInstance = createInstantSearch();
+      const searchBox = createSearchBox({
+        dispose: jest.fn(),
+      });
+
+      const pagination = createPagination({
+        dispose: jest.fn(),
+      });
+
+      instance.addWidgets([searchBox, pagination]);
+
+      instance.init!(
+        createInitOptions({
+          instantSearchInstance,
+        })
+      );
+
+      expect(searchBox.dispose).toHaveBeenCalledTimes(0);
+      expect(pagination.dispose).toHaveBeenCalledTimes(0);
+
+      // Save the Helper to be able to simulate the search
+      const helper = instance.getHelper();
+
+      instance.dispose!(
+        createDisposeOptions({
+          helper: instantSearchInstance.helper!,
+          state: instantSearchInstance.helper!.state,
+        })
+      );
+
+      expect(searchBox.dispose).toHaveBeenCalledTimes(1);
+      expect(pagination.dispose).toHaveBeenCalledTimes(1);
+      expect(pagination.dispose).toHaveBeenCalledWith({
+        state: helper!.state,
+        helper,
+      });
+    });
+
+    it('keeps the widgets on the index', () => {
+      const instance = index({ indexName: 'index_name' });
+      const instantSearchInstance = createInstantSearch();
+      const searchBox = createSearchBox();
+      const pagination = createPagination();
+
+      instance.addWidgets([searchBox, pagination]);
+
+      instance.init!(
+        createInitOptions({
+          instantSearchInstance,
+        })
+      );
+
+      expect(instance.getWidgets()).toHaveLength(2);
+
+      instance.dispose!(createDisposeOptions());
+
+      expect(instance.getWidgets()).toHaveLength(2);
+    });
+
+    it('removes the internal Helper', () => {
+      const instance = index({ indexName: 'index_name' });
+      const instantSearchInstance = createInstantSearch();
+
+      instance.init!(
+        createInitOptions({
+          instantSearchInstance,
+        })
+      );
+
+      expect(instance.getHelper()).toBeDefined();
+
+      instance.dispose!(createDisposeOptions());
+
+      expect(instance.getHelper()).toBe(null);
+    });
+
+    it('removes the DerivedHelper', async () => {
```

Comment:
Is this test doing what it's supposed to do?

Reply:
Yes, it does. We do not expose the `derivedHelper` through the `index` widget. This test case checks that the listeners are correctly removed. It assumes that it's enough to consider the `DerivedHelper` removed. I've updated the test name to be more clear. I also added another one that uses the Helper to checks whether or not the `DerivedHelper` has been removed.